### ISSUE TITLE
`\(::UniformScaling, ::Diagonal{T,Array{T,1}})` returns dense `Matrix` when `T` is abstract

### ISSUE THREAD

**Issue by StevenWhitaker** [JuliaLang/julia#37359]
As stated in the title, `\(::UniformScaling, ::Diagonal{T,Array{T,1}})` returns a dense `Matrix` when `T` is abstract. It seems that, ideally, a `Diagonal` matrix should be returned (as is the case when `T` is concrete). (This is for the case when `T <: Number` is `true`; I haven't looked at what happens when `T <: Number` is `false`.)
```julia
julia> versioninfo()
Julia Version 1.5.1
Commit 697e782ab8 (2020-08-25 20:08 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 8

julia> I \ Diagonal(ones(Integer, 4)) # T abstract --> returns Array
4×4 Array{Float64,2}:
 1.0  0.0  0.0  0.0
 0.0  1.0  0.0  0.0
 0.0  0.0  1.0  0.0
 0.0  0.0  0.0  1.0

julia> I \ Diagonal(ones(Int, 4)) # T concrete --> returns Diagonal
4×4 Diagonal{Float64,Array{Float64,1}}:
 1.0   ⋅    ⋅    ⋅ 
  ⋅   1.0   ⋅    ⋅ 
  ⋅    ⋅   1.0   ⋅ 
  ⋅    ⋅    ⋅   1.0
```

**Comment by martinholters** [comment]
Fix should be as simple as:
```julia
--- a/stdlib/LinearAlgebra/src/diagonal.jl
+++ b/stdlib/LinearAlgebra/src/diagonal.jl
@@ -174,6 +174,7 @@ end
 (*)(x::Number, D::Diagonal) = Diagonal(x * D.diag)
 (*)(D::Diagonal, x::Number) = Diagonal(D.diag * x)
 (/)(D::Diagonal, x::Number) = Diagonal(D.diag / x)
+(\)(x::Number, D::Diagonal) = Diagonal(x \ D.diag)
 
 function (*)(Da::Diagonal, Db::Diagonal)
     nDa, mDb = size(Da, 2), size(Db, 1)
```
If someone wants to try this out and make a PR out of it - go for it!

**Comment by dkarrasch** [comment]
Fixed in #39701.
